File: problems/0200/solution.cpp
#include <vector>
using namespace std;

class Solution {
 private:
  int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
  void gridDFS(vector<vector<char>>& grid, int r, int c) {
    if (r < 0 || r >= grid.size() || c < 0 || c >= grid[0].size() ||
        grid[r][c] == '0') {
      return;
    }

    grid[r][c] = '0';
    for (auto& d : dirs) {
      gridDFS(grid, r + d[0], c + d[1]);
    }
  }

 public:
  int numIslands(vector<vector<char>>& grid) {
    int res = 0;
    for (int i = 0; i < grid.size(); i++) {
      for (int j = 0; j < grid[0].size(); j++) {
        if (grid[i][j] == '1') {
          res++;
          gridDFS(grid, i, j);
        }
      }
    }
    return res;
  }
};

File: problems/0200/solution.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        res = 0

        def grid_dfs(r: int, c: int):
            if (
                r < 0
                or r >= len(grid)
                or c < 0
                or c >= len(grid[0])
                or grid[r][c] == "0"
            ):
                return
            grid[r][c] = "0"

            for dr, dc in dirs:
                grid_dfs(r + dr, c + dc)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    res += 1
                    grid_dfs(i, j)
        return res
